File: python3/day_5_Random Pick with Weight.py
import random

class Solution:

    def __init__(self, w: List[int]):
        self.m = []
        s = 0
        for d in w:
            s += d
            self.m.append(s)

    def pickIndex(self) -> int:
        n = random.randint(1, self.m[-1])
        return bisect.bisect_left(self.m, n)


# Your Solution object will be instantiated and called as such:
# obj = Solution(w)
# param_1 = obj.pickIndex()

File: cpp/day_5_Random Pick with Weight.cpp
class Solution {
    vector<int> m;
    random_device rd;
public:
    Solution(vector<int> w) {
        int sum = 0;
        for (int i = 0; i < w.size(); ++i){
            sum += w[i];
            m.push_back(sum);
        }
    }
    int pickIndex() {
        return lower_bound(m.begin(), m.end(), rd() % m.back() + 1) - m.begin();
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(w);
 * int param_1 = obj->pickIndex();
 */
